### PR TITLE
updating to use new puppetserver_gem provider in place of pe_puppetse…

### PR DESCRIPTION
**Pull request by smbambling**
Use the new puppetserver_gem provider in place of the deprecated pe_puppetserver_gem

### CODE REVIEW

**Comment by ktreese** [comment]
I'm a little confused on how this is supposed to work for Puppet Enterprise, particularly in 2015.2.3.  The provider variable is always being set to "puppet_gem" from the following if condition on line 40 of params.pp:

```
    if versioncmp($::puppetversion, '4.0.0') >= 0 {
      # Configure for AIO packaging.
      $provider = 'puppet_gem'
      $confdir  = '/etc/puppetlabs/code'
      $cmdpath  = ['/opt/puppetlabs/puppet/bin', '/usr/bin', '/usr/local/bin']
    }
```

I want this set to `puppetserver_gem` so that a `puppetserver gem install hiera-eyaml` is performed but in the current logic, this isn't working, nor is a provider override option available (as seems to be the purpose of #90.  Am I missing something?

**Comment by smbambling** [comment]
Ah, your totally correct...your PR is also required. to make this work correctly.

**Comment by smbambling** [comment]
I also think wonder if pe_version got renamed to pe_build

```
pe_build => 2015.2.2
```

**Comment by ktreese** [comment]
Yes -- it gets confusing because in 2015, is_pe was changed from false to true.  https://tickets.puppetlabs.com/browse/MODULES-2279

Because of this, line 17 of the params manifest is false, therefore the provider is never set to puppetserver_gem on line 26, and the else condition is met, setting provider = puppet_gem on line 42.  I'm not sure if this was the intention of the author whereby PE users must override provider to puppetserver_gem, or if the logic is supposed to detect PE vs nonPE installations.

**Comment by ktreese** [comment]
Based on the values of `$hiera_yaml` and `$data_dir` on line 18 and 19 respectively, it would seem this was the intention.

**Comment by smbambling** [comment]
oh this logic is totally broke then....bummer.  I'm working on this now including your fix from #90

**Comment by smbambling** [comment]
@ktreese ok i'm made some changes and just tested by building a PE 2015.2 box from scratch 

```
[root@mom ~]# /opt/puppetlabs/bin/puppetserver gem list

*** LOCAL GEMS ***

deep_merge (1.0.1)
hiera-eyaml (2.0.8)
hiera-eyaml-gpg (0.6)
highline (1.6.21)
jar-dependencies (0.1.13)
jruby-openssl (0.9.7 java)
json (1.8.0 java)
puppetclassify (0.1.2)
rake (10.1.0)
rdoc (4.1.2)
ruby_gpg (0.3.2)
trollop (2.1.2)
[root@mom ~]# /opt/puppetlabs/puppet/bin/gem list

*** LOCAL GEMS ***

bigdecimal (1.2.4)
colored (1.2)
cri (2.6.1)
deep_merge (1.0.1)
faraday (0.9.1)
faraday_middleware (0.9.1)
faraday_middleware-multi_json (0.0.6)
hiera-eyaml (2.0.8)
highline (1.6.21)
hocon (0.9.3)
io-console (0.4.3)
json (1.8.1)
json_pure (1.8.2)
log4r (1.1.10)
minitar (0.5.4)
minitest (4.7.5)
multi_json (1.11.0)
multipart-post (2.0.0)
net-ssh (2.1.4)
psych (2.0.5)
puppetclassify (0.1.2)
r10k (2.0.3)
rake (10.1.0)
rdoc (4.1.0)
rugged (0.21.4)
semantic_puppet (0.1.0)
stomp (1.3.3)
test-unit (2.1.6.0)
trollop (2.1.2)
```

**Comment by smbambling** [comment]
I still think this needs more cleanup ..but this makes it functional

**Comment by ktreese** [comment]
@smbambling This works, however the confdir is now set incorrectly since `pe_server_version` will always be satisfied in 2015, whereby `confdir` is set to the old /etc/puppetlabs/puppet.  Because of this, hiera.yaml is being stored here, but the correct location is /etc/puppetlabs/code per https://docs.puppetlabs.com/puppet/4.0/reference/whered_it_go.html#new-codedir-holds-all-modulesmanifestsdata

I think the first if statement to check `is_pe` is for earlier versions of PE, and the `$::pe_server_version` is for PE >= 4.0.0.  Since the provider is the pe_server_version conditional is not set to use puppetserver, I assume it must be set via an override.  Perhaps @hunner can weigh in on whether or not the module is supposed to determine PE vs OSS, or if we're supposed to throw in an override.  For now, I forked your repo and added/merged in #90 into the puppetserver_gem_provider branch.  I then pinned my Puppetfile to the puppetserver_gem_provider branch of my fork, and this works, at least in the way I expect it to.

**Comment by smbambling** [comment]
@ktreese I 'think' I have the logic correct...though its getting pretty dirty and looks like this should be cleaned up in the future.  But I've tested that it correctly the the confdir (coded) for PE 2015.2 or higher and uses /etc/puppetlab/puppet for legacy.  I believe this maintains backwards compatibility.

**Comment by smbambling** [comment]
@ktreese I've also added your change from PR #90

**Comment by ktreese** [comment]
@smbambling, this works correctly in my tests as well.  Just to clarify and give credit where it's due, PR #90 is from @jyaworski :smile:

**Comment by smbambling** [comment]
@ktreese @jyaworski @hunner I did some cleanup of the logic in the params.pp and re-tested PE 2015.2.2 and 2015.3 

If anyone has a FOSS puppet master to test on that would be good

**Comment by smbambling** [comment]
Also updated the params.pp to fix the issues with running the spec tests.

```
 bundle exec rake spec
```

**Comment by jyaworski** [comment]
@smbambling could you rebase? Tag @hunner

**Comment by smbambling** [comment]
@jyaworski: this branch is rebased off the latest upstream/master

**Comment by jyaworski** [comment]
Sorry; I use rebase and squash interchangeably when I probably shouldn't. I _meant_ squash. Thanks.

**Comment by hunner** [comment]
Thanks for this! I totally missed it somehow :(

**Comment by smbambling** [comment]
No worries let me know if there is any recommendation for modifications.

File: manifests/eyaml.pp [hunk 1]
```diff
@@ -25,7 +25,7 @@
     undef   => 'installed',
     default => $eyaml_version,
   }
-  if $provider == 'pe_puppetserver_gem' {
```
Comment (hunner) [inline manifests/eyaml.pp]: It still needs to handle pe_puppetserver_gem for PE versions older than 2015.2

File: manifests/params.pp [hunk 2]
```diff
@@ -14,45 +14,44 @@
...
-    if $::pe_server_version {
-      $master_service = 'pe-puppetserver'
-    } else {
-      $master_service = 'puppetmaster'
-    }
-    if $::puppetversion and versioncmp($::puppetversion, '4.0.0') >= 0 {
+    $owner    = 'puppet'
+    $group    = 'puppet'
+    $master_service = 'puppetmaster'
+    $hiera_yaml = "${confdir}/hiera.yaml"
```
Comment (hunner) [inline manifests/params.pp]: $confdir isn't set at this point.

File: metadata.json [hunk 3]
```diff
@@ -51,6 +51,10 @@
     {
       "name": "puppetlabs-stdlib",
       "version_requirement": ">= 4.3.1 < 5.0.0"
+    },
+    {
+      "name": "puppetlabs-puppetserver_gem",
+      "version_requirement": ">= 0.1.0"
```
Comment (hunner) [inline metadata.json]: I wouldn't add a hand dependency here, since anyone running masterless setups wouldn't need this module.

**Comment by hunner** [comment]
When reading through this and following up on a few other bugs, I ran across lots of things that needed fixing on 2015.x; I've ported your changes and other fixes to #102 which appears to need a major 2.0.0 release to do it right, unfortunately.

**Comment by smbambling** [comment]
Cool, I'll review #102 I noticed you talked about having issue with hiera.yaml in the code dir.  Thats def something I ran into I tweaked the the hiera.yaml location so that code manager would still work correctly.


Seems like having it live in puppet/hiera.yaml is the 'best' place for it and a sane default.

```
 hiera_yaml           => '/etc/puppetlabs/puppet/hiera.yaml',
```

**Comment by jyaworski** [comment]
Hello:

I can confirm that I've needed to place my hiera config inside /etc/puppetlabs/puppet as well.

**Comment by hunner** [comment]
@jyaworski @smbambling Thanks for that feedback! So now I'm trying to decide if the hiera module should be responsible for updating the `hiera_config` setting in puppet.conf or not, and if it should use augeas or ini_setting. What do you think? Is updating puppet.conf out of scope for the module? If it doesn't would it confuse new users?

**Comment by jyaworski** [comment]
@hunner there is precedent for it in the puppetdb module with storeconfigs, and I really dislike it. Maybe instead of updating puppet.conf we could do a symlink like r10k does.

**Comment by smbambling** [comment]
@hunner: I currently use the PuppetLabs [infile](https://github.com/puppetlabs/puppetlabs-inifile) module to set this setting in my profile for our Puppet Masters.

```
ini_setting { 'puppetserver conf hiera_config':
    ensure  => present,
    path    => '/etc/puppetlabs/puppet/puppet.conf',
    section => 'main',
    setting => 'hiera_config',
    value   => '/etc/puppetlabs/puppet/hiera.yaml',
  }
```

I think that having this module set that configuration for hiera_config is still in scope.  Trying to think of a way to verify if the parameter is overridden from the 'default' value from params.pp and if that setting is modified then also use ini_settings to set the hiera_config option.

Another option might be to just have a parameter that would allow users to enable or disable the module from setting the hiera_config in puppet.conf.

@jyaworski: To my understanding the symlink that is created in /etc is used for the command line utility and some backwards compatibly.  Not sure its as intuitive to use a symlink as it would be to look in the puppet.conf config.

**Comment by npwalker** [comment]
+1 to the idea of managing hiera_config in puppet.conf by default but allowing it to be disabled by a parameter if someone would prefer that it not be managed.

**Comment by hunner** [comment]
Thanks! I'll go with using ini_setting to update puppet.conf with a parameter to disable that. I think creating a symlink in /etc/puppetlabs/code/hiera.yaml would run into the same issue that putting the file there did.

**Comment by smbambling** [comment]
Yeah the symlink would still have the issue with code manager in 2015.3 over writing the file

**Comment by smbambling** [comment]
This logic from PR #102 and the fix from #110 Look to cover all the bases for this module. Uutside setting the hard dependency for puppetserver_gem, which @hunner correctly pointed out would not be needed on a masterless setup

**Pull request closed without merging**